src/babel-serializer/geometry/stl.cc

```diff
@@ -0,0 +1,272 @@
+            if (pos + count > data.size())
+                on_error(data, data.first(last_pos), "Failed to parse binary stl-file: unexpected eof", severity::error);
+            pos += count;
+            return data.subspan(last_pos, count);
+        };
+
+        stl::geometry::color default_color = {std::byte(0.6 * 255), std::byte(0.6 * 255), std::byte(0.6 * 255), std::byte(255)};
+
+        auto const header = get_next(80);
+        // check if this is Materialise Magics software style
+        auto is_materialize = false;
+        // the 10 comes from the bytes "COLOR=" and 4 RGBA bytes
+        for (size_t i = 0; i < header.size() - 10; ++i)
+        {
+            auto const sv = header.subspan(i, 10);
+            auto const s = cc::string_view(reinterpret_cast<char const*>(sv.begin()), reinterpret_cast<char const*>(sv.end()));
+            if (s.starts_with("COLOR="))
+            {
+                is_materialize = true;
+                default_color.r = sv[6];
+                default_color.g = sv[7];
+                default_color.b = sv[8];
+                default_color.a = sv[9];
+                break;
+            }
+        }
+
+        auto const parse_uint32 = [&]() {
+            cc::uint32 ui;
+            auto const s = get_next(sizeof(cc::uint32));
+            std::memcpy(&ui, s.data(), sizeof(ui));
+            return ui;
+        };
+
+        auto const parse_uint16 = [&]() {
+            cc::uint16 ui;
+            auto const s = get_next(sizeof(cc::uint16));
+            std::memcpy(&ui, s.data(), sizeof(ui));
+            return ui;
+        };
+
+        auto const parse_float = [&]() {
+            cc::float32 f;
+            auto const s = get_next(sizeof(cc::float32));
+            std::memcpy(&f, s.data(), sizeof(f));
+            return f;
+        };
+
+        auto const n_triangles = parse_uint32();
+
+        if (data.size() < 84 + n_triangles * 50)
+        {
+            on_error(data, data, "Failed to parse binary STL-file: File too short", severity::error);
+            return;
+        }
+
+        geometry.triangles.resize(n_triangles);
+        geometry.triangle_color.resize(n_triangles);
+        for (size_t i = 0; i < n_triangles; ++i)
+        {
+            auto& t = geometry.triangles[i];
+            t.normal.x = parse_float();
+            t.normal.y = parse_float();
+            t.normal.z = parse_float();
+            t.v0.x = parse_float();
+            t.v0.y = parse_float();
+            t.v0.z = parse_float();
+            t.v1.x = parse_float();
+            t.v1.y = parse_float();
+            t.v1.z = parse_float();
+            t.v2.x = parse_float();
+            t.v2.y = parse_float();
+            t.v2.z = parse_float();
+
+            // this is the "attribute_byte_count" but in practice is always a color
+            // see https://en.wikipedia.org/wiki/STL_(file_format)#Color%20in%20binary%20STL
+            auto const color = parse_uint16();
```

Comment:
this info would also be useful in the geometry struct definition

Reply:
It technically is, `We accept colors in VisCAM and SolidView, as well as Materialise Magics format`, however I can see that it is not obvious that this relates to color only.

Reply:
oh nevermind, i though the color values themselves are still encoded in the output, but they're just intensities in [0, 255] right? in that case the type should probably be uint8

Reply:
Yeah, I was unsure about the type. We could also do `tg::color<3, uint8>`, however I am unsure about the state of `tg::color` with non-float types.

Reply:
the simple struct is fine i think, but byte signals "opaque memory" while it really is an integer

Reply:
uint8_t is better for now. proper color management in tg is on my list though